org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/ChainCompletionProposalComputer.java

```diff
@@ -263,31 +251,86 @@ private boolean matchesExpectedPrefix(final IJavaElement element) {
 		return String.valueOf(element.getElementName()).startsWith(prefix);
 	}
 
-	private CompletionItem create(final Chain chain) throws JavaModelException {
-		final String insert = createInsertText(chain, chain.getExpectedDimensions());
-		final CompletionItem ci = new CompletionItem();
+	// given Collection.emptyList()	return Collection.emptyList
+	// args[i].lines().toList()	return args[i].lines().toList
+	private char[] getQualifiedMethodName(String name) {
+		var nonSnippetName = removeSnippetString(name);
+		int index = nonSnippetName.lastIndexOf('(');
+		if (index > 0) {
+			return nonSnippetName.substring(0, index).toCharArray();
+		}
+		return nonSnippetName.toCharArray();
+	}
 
-		ci.setTextEditText(insert);
-		ci.setInsertText(getQualifiedMethodName(insert));
-		ci.setInsertTextFormat(snippetStringSupported ? InsertTextFormat.Snippet : InsertTextFormat.PlainText);
-		ci.setKind(CompletionItemKind.Method);
-		setLabelDetails(chain, ci);
+	private String removeSnippetString(String name) {
+		return name.replaceAll("\\[\\$\\{.*\\}\\]", "[i]");
+	}
```

Comment:
Is this actually necessary ? It looks like only `appendParameters` and `appendArrayDimensons` were creating snippets (placeholders) but you've removed those, so this shouldn't be necessary.

Reply:
`final var insertText = createInsertText(chain, chain.getExpectedDimensions());` still could have snippet string isn't it based on the value of 'this.snippetStringSupported` at `appendParameters` ? 

`insertText` is used to compute the qualified name by passing it into `getQualifiedMethodName`. or did I misunderstood your question ?

Reply:
Yes, but in your change, [you removed the part in `appendParameters` that would insert the snippet string](https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/2835/files#diff-6dc3eaaa7bd1b0c7ad2bcf6cab295d1339ac417c98ebb0aa4424b7bcd8c362d1L390). So I think even the `snippetStringSupported` part isn't needed anymore.

If I understand correctly, by choosing to deal with CompletionProposals everywhere, you've basically pushed the actual introduction of snippets to https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/96ca6b632e3aa46260a52da4de4e6d100ab899e4/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/CompletionProposalRequestor.java#L419 , which eventually ends up in https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/96ca6b632e3aa46260a52da4de4e6d100ab899e4/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/CompletionProposalReplacementProvider.java#L625, which is where they happen for all other completions.

Basically I don't see a case where you need to remove the snippet string from insertText, because there should not be anything added yet. If i'm missing something let me know.